### ISSUE TITLE
[FR]: Optimizer - Improve Optimisation with a Test/Train Dataset split

### ISSUE THREAD

**Issue by FlyingGorillaz** [comet-ml/opik#4080]
### Problem Statement
The Opik optimizer currently lacks built-in support for train/test dataset splitting during prompt optimization. 

As documented in the Optimizer Datasets & Testing guide:

"The opik_optimizer itself does not enforce a strict train/validation/test split within the optimize_prompt call in the same way some traditional ML model training frameworks do. You are responsible for managing your datasets appropriately for robust evaluation."

This limitation creates several critical issues:

**Data Pollution Risk:** When the same dataset is used for both optimization and final evaluation, optimizers can inadvertently overfit to specific test examples, leading to inflated performance metrics that don't generalize.

No Ground Truth for Generalization: Without a held-out test set, users cannot reliably assess whether their optimized prompts will perform well on unseen data in production.

**Manual Implementation Burden:** Users must manually split datasets before calling optimize_prompt(), creating inconsistency across implementations and increasing the likelihood of mistakes.

### Motivation

**Misalignment with ML Best Practices:** Standard machine learning workflows always separate training and evaluation data. The current design violates this fundamental principle.
The GEPA optimizer documentation acknowledges this limitation:

"_The GEPA Score column reflects the aggregate score GEPA computes on its train/validation split when deciding which candidates stay on the Pareto front... The Opik Score column is a fresh evaluation performed through Opik's metric pipeline on the same dataset."_

This shows that some optimizers (GEPA) have internal train/validation splits, but this isn't standardized across all optimizers, and critically, no held-out test set is preserved for a final unbiased evaluation. 

**Often the optimized prompt includes training data as a few-shot prompt, making the improved system prompt better for the dataset but not necessarily delivering improved performance in real world tests.**

**Comment by FlyingGorillaz** [comment]
Not that some of the optimisation techniques attempt test/train splits, but do so erroneously. e.g. https://github.com/comet-ml/opik/blob/main/sdks/opik_optimizer/src/opik_optimizer/few_shot_bayesian_optimizer/few_shot_bayesian_optimizer.py

Uses whatever is in the Dataset you pass to optimize_prompt.
The few shot system prompt template is built from up to 10 items from the FULL dataset. That is to say that the Few shot examples per trial are sampled from ALL items in the entire dataset.

There is no concept of "training data" vs "test data" within this class. If the Dataset object contains both train and test items, this optimiser will happily:

1) Use test items as examples
2) Score/Eval on those same items.

So if I want a clean test/train data set, I currently need to enforce it myself, for example:
1) Pass a training dataset that contains only training items to optimize_prompt
2) Keep a separate held-out dataset that I never pass into this optimiser, and use that only for a separate final evaluation.

Without this, the Eval scores are completely unreliable as the "best" system prompt contains the precise prompts from the test dataset.

It looks like may come from this area of the code not using the training dataset, but instead using the full dataset.

Create the few-shot prompt template
        with reporting.creation_few_shot_prompt_template(
            verbose=self.verbose
        ) as fewshot_template_report:
            fewshot_template = self._create_fewshot_prompt_template(
                model=self.model,
                prompt=prompt,
                few_shot_examples=[
                    {k: v for k, v in item.items() if k != "id"}
                    for item in dataset.get_items(nb_samples=10)
                ],
            )

**Comment by jverre** [comment]
Hi @FlyingGorillaz 

This makes sense, we will take a look at adding this - Would be nice for it to be included in the library.

In the meantime you can define two datasets, one train and one test. You would then run the optimization on the train dataset and the final evaluation on the test dataset which would mimic the train / test split you described above
